### PR TITLE
ci: reenable CI on GitHub Actions

### LINKED ISSUE #16: Create a GitHub Actions continuous integration workflow.

### PR DESCRIPTION
**Pull request by deliaBlue**
closes #16

### CODE REVIEW

**Comment by uniqueg** [comment]
The test is currently failing, so no need to request my review at this point. Please make sure the test is working, then re-request a review. If you cannot solve the problem on your own, describe here what you tried and I will help you fix it.

**Comment by deliaBlue** [comment]
After finding why the test fails and fix it, now the test fails as it has no files to run the test with. Just for this PR I would add to the branch the test files and after checking the test is correct and misses nothing I'd delete the test files and address the issue #17 .

**Comment by uniqueg** [comment]
> After finding why the test fails and fix it, now the test fails as it has no files to run the test with. I will assign this branch to issue #17 (create test files for GitHub Actions) along with issue #10 (Change chromosome on test files) too and create the test files.

Or: You remove the "Run local test" step from the CI workflow, undo all other changes, and then we can already merge that. At least it tests against syntax errors in the SMK descriptor files (the rule graph wouldn't be built otherwise!) and against build issues (issues with Conda etc.). So I think it's a useful pipeline to build on. Then tackle #10 and #17 in the next PR and re-add the "Run local test" step...

**Comment by deliaBlue** [comment]
> > After finding why the test fails and fix it, now the test fails as it has no files to run the test with. I will assign this branch to issue #17 (create test files for GitHub Actions) along with issue #10 (Change chromosome on test files) too and create the test files.
> 
> Or: You remove the "Run local test" step from the CI workflow, undo all other changes, and then we can already merge that. At least it tests against syntax errors in the SMK descriptor files (the rule graph wouldn't be built otherwise!) and against build issues (issues with Conda etc.). So I think it's a useful pipeline to build on. Then tackle #10 and #17 in the next PR and re-add the "Run local test" step...

Oh! did not see the message on time (nor on splitting this PR instead of merging two more issues on it) and made a new commit... If the CI workflow fails again I'll do it as you say. Also, I've a small report on how I split the reference genome, change the genomic regions in the annotation files to meet these new coords and how the test library is build. Should I post it somewhere or it is not necessary?

**Comment by deliaBlue** [comment]
A couple of comments:

1. Locating Singularity images takes ~24' out of the 30' it takes the "Run local test" to complete". I'm not sure if that's normal or not but when running the test locally it does not take that long, never.
2. Now the test fails on "Run test for rule graph". The error states that the "dot" command is not found. When trying the CI on a private repo, this problem was fixed by adding `graphviz=7.1.0` to the dependencies but I saw on the `zarp` repo they have the same command but not the dependency (on any `yml` file). I do not know if there is another way to fix it, otherwise, I'll proceed on adding `graphviz=7.1.0` to the `environment.yml`.

**Comment by uniqueg** [comment]
2. Maybe one of the other dependencies in ZARP requires `graphviz` and that's why it's listed explicitly, dunno. If you need it, add it, no problem.

1. There are a lot of images that need to be pulled and then converted. 24 min seems excessive, but perhaps Docker or GH servers are slow, or they are throttled (we are not paying them a dime after all, and these things are expensive). Also, in local execution, images would be cached. Same for our test VM (which is one reason why we ran out of space). There is, I think, some caching on GH Actions, too, but I don't know how it works. Maybe things will improve over time. You can also try to search online and see if there's something you can do about (image) caching in GitHub Actions.

One thing you can do is split long-running from short-running tests, if possible. I would still only merge when _all_ of the tests succeed, but perhaps at least during development it could help. But not sure if it makes sense here.

In general, long-running tests can be a pain. Apart from doing what you can to keep your tests short, best is probably to work on different features or projects at the same time. Like the people in the lab have incubation times where they need to wait for some reaction. Guess it's part of the job 🤷‍♂️

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          auto-activate-base: false
+      
+      - name: Update mirflowz env with root packages
+        run: mamba env update -n mirflowz -f environment.root.yml
+        
+      - name: Run local test
+        run: bash test/test_workflow_local.sh
+      
+      - name: Run test for rule graph
+        run: bash test/test_rule_graph.sh
```
Comment (uniqueg) [inline .github/workflows/CI.yml:44]: I would put that before running the test script, as it's faster

**Review by uniqueg — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      -  name: Checkout Repository
+         uses: actions/checkout@v3
+         
+      - name: Setup environment
+        uses: conda-incubator/setup-miniconda@v2
+        with:
+          python-version: "3.9.16"
+          mamba-version: "*"
+          channels: conda-forge
+          channel-priority: true
```
[the remark this answers is not in the document]
Comment (deliaBlue) [reply .github/workflows/CI.yml:32]: Except for the `mamba-version` the other options seemed to not required.

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    branches:
+      - '*'
+ 
+jobs:
+
+  snakemake-test:
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: bash -l {0}
```
[the remark this answers is not in the document]
Comment (deliaBlue) [reply .github/workflows/CI.yml:20]: As seen in [stack overflow](https://stackoverflow.com/questions/69070754/shell-bash-l-0-in-github-actions), the `-l` option gives a _login_ bash in which to set the environment and the `{0}` is a template place holder,  replaced at pipeline execution time by the actual script command to execute. 

When removed the test fails and gives an exit status 127; it does not find the command "snakemake" which makes me think that the environment is not set.

File: test/test_files/GRCh38_UCSC2ensembl_19.txt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+chr19_KI270920v1_alt	
+chr19_KI270921v1_alt	
+chr19_KI270922v1_alt	
+chr19_KI270923v1_alt	
+chr19_KI270929v1_alt	
+chr19_KI270930v1_alt	
+chr19_KI270931v1_alt	
+chr19_KI270932v1_alt	
+chr19_KI270933v1_alt	
+chr19_KI270938v1_alt
```
[the remark this answers is not in the document]
Comment (deliaBlue) [reply test/test_files/GRCh38_UCSC2ensembl_19.txt:44]: When removing all these entries the output turns out to be another. I could not figure out why. I'm changing too the `expected_output.md` to meet the new results.

**Review by uniqueg — APPROVED** [review]

File: .github/workflows/CI.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (uniqueg) [reply .github/workflows/CI.yml:17]: Thanks!

**Pull request merged**